quic/congestion_control/Copa2.cpp

```diff
@@ -0,0 +1,262 @@
+    lossCycleStartTime_ = Clock::now();
+    return;
+  }
+
+  auto numPktsInLossCycle = numAckedInLossCycle_ + numLostInLossCycle_;
+  if (*sentTime < lossCycleStartTime_) {
+    // Wait for at-least one RTT before declaring lossyMode_
+    return;
+  }
+  if (numPktsInLossCycle < 2 / lossToleranceParam_ && numLostInLossCycle_ < 2) {
+    // Second condition is needed in case there are losses, but not acks
+    return;
+  }
+  VLOG(5) << __func__ << " lossyMode=" << lossyMode_
+          << " num lost=" << numLostInLossCycle_
+          << " num acked=" << numAckedInLossCycle_ << " " << conn_;
+  // Cycle has ended. Take stock of the situation
+  DCHECK(numPktsInLossCycle > 0);
+  lossyMode_ = numLostInLossCycle_ >= numPktsInLossCycle * lossToleranceParam_;
+  numAckedInLossCycle_ = 0;
+  numLostInLossCycle_ = 0;
+  lossCycleStartTime_ = Clock::now();
+}
+
+void Copa2::onPacketLoss(const LossEvent& loss) {
+  VLOG(10) << __func__ << " lostBytes=" << loss.lostBytes
+           << " lostPackets=" << loss.lostPackets << " cwnd=" << cwndBytes_
+           << " inflight=" << conn_.lossState.inflightBytes << " " << conn_;
+  if (conn_.qLogger) {
+    conn_.qLogger->addCongestionMetricUpdate(
+        conn_.lossState.inflightBytes,
+        getCongestionWindow(),
+        kCongestionPacketLoss);
+  }
+  DCHECK(loss.largestLostPacketNum.has_value());
+  subtractAndCheckUnderflow(conn_.lossState.inflightBytes, loss.lostBytes);
+  if (loss.persistentCongestion) {
+    VLOG(10) << __func__ << " writable=" << getWritableBytes()
+             << " cwnd=" << cwndBytes_
+             << " inflight=" << conn_.lossState.inflightBytes << " " << conn_;
+    cwndBytes_ = conn_.transportSettings.minCwndInMss * conn_.udpSendPacketLen;
+    if (conn_.pacer) {
+      // TODO Which min RTT should we use?
+      conn_.pacer->refreshPacingRate(cwndBytes_, conn_.lossState.mrtt);
+    }
+    if (conn_.qLogger) {
+      conn_.qLogger->addCongestionMetricUpdate(
+          conn_.lossState.inflightBytes,
+          getCongestionWindow(),
+          kPersistentCongestion);
+    }
+  }
+
+  numLostInLossCycle_ += loss.lostPackets;
+  manageLossyMode(loss.largestLostSentTime);
+}
+
+void Copa2::onPacketAcked(const AckEvent& ack) {
+  DCHECK(ack.largestAckedPacket.has_value());
+  subtractAndCheckUnderflow(conn_.lossState.inflightBytes, ack.ackedBytes);
+  minRTTFilter_.Update(
+      conn_.lossState.lrtt,
+      std::chrono::duration_cast<microseconds>(ack.ackTime.time_since_epoch())
+          .count());
+  bytesAckedInCycle_ += ack.ackedBytes;
+  for (const auto& ackPkt : ack.ackedPackets) {
+    appLimitedInCycle_ = appLimitedInCycle_ || ackPkt.isAppLimited;
+  }
+  auto rttMin = minRTTFilter_.GetBest();
+
+  numAckedInLossCycle_ += ack.ackedPackets.size();
+  manageLossyMode(ack.largestAckedPacketSentTime);
+
+  auto dParam = rttMin;
+  if (lossyMode_) {
+    dParam = duration_cast<microseconds>(rttMin * 2. * lossToleranceParam_);
+  }
+  auto cycleDur = rttMin + dParam;
```

Comment:
Would love to have some comments on what this value means, and what role it plays

Reply:
Added comments. It makes Copa2 less aggressive when there is excessive loss since that indicates that there is a short buffer